src/passkey_child/passkey_child_common.c

```diff
@@ -319,3 +376,181 @@ register_key(struct passkey_data *data)
+        DEBUG(SSSDBG_OP_FAILURE, "d2i_pubkey failed [%lu][%s].\n",
+              err, ERR_error_string(err, NULL));
+        ret = EIO;
+        goto done;
+    }
+
+    base_id = EVP_PKEY_get_base_id(evp_pkey);
+    if (base_id == EVP_PKEY_EC) {
+        _pk_data->type = COSE_ES256;
+        ret = evp_pkey_to_es256_pubkey(evp_pkey, _pk_data);
+    } else if (base_id == EVP_PKEY_RSA) {
+        _pk_data->type = COSE_RS256;
+        ret = evp_pkey_to_rs256_pubkey(evp_pkey, _pk_data);
+    } else if (base_id == EVP_PKEY_ED25519) {
+        _pk_data->type = COSE_EDDSA;
+        ret = evp_pkey_to_eddsa_pubkey(evp_pkey, _pk_data);
+    } else {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unrecognized key type.\n");
+        ret = EINVAL;
+    }
+    if (ret != EOK) {
+        goto done;
+    }
+
+    ret = EOK;
+
+done:
+    if (evp_pkey != NULL) {
+        EVP_PKEY_free(discard_const(evp_pkey));
+    }
+    talloc_free(tmp_ctx);
+
+    return ret;
+}
+
+errno_t
+authenticate(struct passkey_data *data)
+{
+    fido_assert_t *assert = NULL;
+    fido_dev_info_t *dev_list = NULL;
+    fido_dev_t *dev = NULL;
+    struct pk_data_t pk_data_t;
+    size_t dev_list_len = 0;
+    int count = 0;
+    errno_t ret;
+
+    dev_list = fido_dev_info_new(DEVLIST_SIZE);
+    if (dev_list == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_info_new failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Checking for devices.\n");
+    ret = list_devices(dev_list, &dev_list_len);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "%d key handles provided.\n",
+          data->keys_size);
+
+    while (count < data->keys_size) {
+        DEBUG(SSSDBG_TRACE_FUNC,
+              "Preparing assert request data with key handle %d.\n", count + 1);
+
+        assert = fido_assert_new();
+        if (assert == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_assert_new failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        DEBUG(SSSDBG_TRACE_FUNC, "Preparing assert request data.\n");
+        ret = prepare_assert(data, count, assert);
+        if (ret != FIDO_OK) {
+            goto done;
+        }
+
+        DEBUG(SSSDBG_TRACE_FUNC, "Selecting device.\n");
+        ret = select_device(dev_list, dev_list_len, assert, &dev);
+        if (ret == EOK) {
+            /* Key handle found in device */
+            break;
+        }
+
+        if (dev != NULL) {
+            fido_dev_close(dev);
+        }
+        fido_dev_free(&dev);
+        fido_assert_free(&assert);
+        count++;
+    }
+
+    if (dev == NULL) {
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Comparing the device and policy options.\n");
+    ret = get_device_options(dev, data);
+    if (ret != FIDO_OK) {
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Resetting assert options.\n");
+    ret = set_assert_options(FIDO_OPT_OMIT, data->user_verification, assert);
```

Comment:
You probably want to set UP to `FIDO_OPT_TRUE` here. Otherwise, `fido_assert_verify()` will not verify that the user presence flag was set in the returned authenticator data.

Reply:
Done